### ISSUE TITLE
Why not use the common  metrics  for Next POI Recommendation?

### ISSUE THREAD

**Issue by kevin-xuan** [Yifang-Qin/DisenPOI#4]
Hello, thanks for your source code. As the title describes, i'm curious about the metrics in your paper.
This work is about Next POI recommendation, but the two metrics: AUC and Logloss, are used. In general, the HR, NDCG, MRR metrics are commonly used in most existing studies. I want to know what will happen when using HR and NDCG?
Therefore, i try to adapt this code as follows:
`
self.MLP = nn.Sequential(
                nn.Linear(4 * self.embed_dim, self.embed_dim),
                nn.LeakyReLU(),
                nn.Linear(self.embed_dim, self.embed_dim)
            ) &
logits_ = self.MLP(torch.cat((tar_embed, tar_geo_embed, sess_enc_p, geo_enc_p), dim=-1))
        fin_logits = logits_.matmul(self.poi_embed.transpose(0,1))
        logits = fin_logits.gather(1, tar_poi.unsqueeze(-1)).squeeze()
`
That is, the logits_'s shape will be [batch_size, embed_dim] instead of [batch_size], fin_logits' shape will be [batch_size, n_poi], and logits (shape: [batch_size]) will be consistent with the original code and could be used to calculate the BCE loss.

However, you mentioned that "For each positive sample, we randomly select one of the user’s unvisited POIs as the negative sample." In this way, we will have two instances (uid, neg_list[i], pos_list[: i], loc_dict[neg_list[i]], 0) and (uid, pos_list[i], pos_list[: i], loc_dict[pos_list[i]], 1). But HR , MRR, and NDCG only need to compare the rank of positives (target POI) in the evaluation phase, so should we  only consider one instance (uid, pos_list[i], pos_list[: i], loc_dict[pos_list[i]], 1) for the val and test set? 
I don't know whether the two modifications are correct? Could you provide some suggestions?

**Comment by kevin-xuan** [comment]
The results are poor.
![截屏2024-02-12 20 45 45](https://github.com/Yifang-Qin/DisenPOI/assets/87954969/c2e03b3b-0b53-49f8-9a25-8eacc76be3e7)
Click Through Rate (CTR) prediction is essentially a binary classification task,  so AUC and logloss is proper. However, according to your problem definition, "Given user 𝑢 and target POI 𝑣, the goal of location-based CTR prediction is to predict the probability for 𝑢 to visit 𝑣", this definition is similar as that of next poi recommendation, i.e., "Given user 𝑢, the goal of next poi recommendation is to predict the target poi 𝑣 for 𝑢 to visit next".
Could you please tell the difference between CTR (binary-class) and next poi recommendation (n_poi-class)?

**Comment by Yifang-Qin** [comment]
Hello Kevin, thanks for your concerns about the evaluation metric of DisenPOI. 

For your question about CTR prediction, as you have mentioned, we didn't adopt other commonly used metrics in sequential recommendations. The fundamental reason of this is because DisenPOI was initially developed for Meituan's advertising system to deal with the **specific needs of making CTR predictions**. Unlike ranking models that recommend a list of items, CTR models focus on binary classification: deciding whether a user will click on an item among a limited set of candidates. As a result, the evaluation of CTR models are more likely to be evaluation of binary classification models. This means we need to preprocess the whole interaction dataset to create some negative samples (by random sampling) and positive samples (observed interactions) with labels. Metrics like AUC and LogLoss are adopted for the same reason. For more details of CTR prediction you can refer to the [DSTN paper](https://arxiv.org/pdf/1906.03776v2.pdf).

For your experimental results, since the original model is optimized by the binary classification target, it suffers when directaly evaluated by ranking objectives like nDCG. To apply DisenPOI to more generalized POI prediction taks, the random negative samples in the proprocessing should be discarded, exactly as you have mentioned. You may also need to change the optimization target to cross-entropy (CE) or other ranking loss functions like BPR loss. It is noteworthy that the optimization target is also crucial to obtain reasonable results in next POI recommendation and I would recommend you try out CE as an early attempt. Additionally, according to my personal experience, using inner product instead of concatenated MLP to generate the final recommend logits also helps improve the performance.

Thanks again for your interests in our work and if you meet any other difficulties in your experiments, feel free to ask me and I'll try my best to answer your questions.

**Comment by kevin-xuan** [comment]
Thanks for your prompt reply and valuable suggestions.

**Comment by kevin-xuan** [comment]
As you suggested, i adapt this code from three aspects as follows:
1. i discard random negative samples in the proprocessing, i.e., only one instance (uid, pos_list[i], pos_list[: i], loc_dict[pos_list[i]], 1) for training, valid, and test set.
2. The original self.MLP maps concatenated representation into a 1-dimensional logits. Instead, self.MLP is modified to map concatenated representation into a d-dimensional logits and then i use inner product between logits and initial POI embeddings to calculate user preference scores (As presented in above code).
3. I also change the optimization target to cross-entropy (CE).
![截屏2024-02-13 09 27 25](https://github.com/Yifang-Qin/DisenPOI/assets/87954969/449f8148-3cb9-4b74-a0dd-aefcc938c5c5)
The results are amazing (0.9999) !!! The reason may be the concatenated representation contains the information of target POI 𝑣 and the similarity calculation leads to a very very high score for target POI. This is unnormal, so i guess the method may be not suitable for more generalized sequential recommendation due to the possible information leakage caused by the discard of negative samples? What do you think？

**Comment by Yifang-Qin** [comment]
You are right about the information leakage issue. When adapting DisenPOI as a regular POI recommendation baseline, the calculation of recommendation logits should exclude any information related to the ground-truth target POI itself.

**Comment by kevin-xuan** [comment]
OK, i see. Thanks for your reply sincerely.
